### PR TITLE
#37 Event visit request

### PR DESCRIPTION
**Pull request by denys-medynskyi**

### CODE REVIEW

File: app/controllers/admin/visit_request/cancel_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    class CancelController < BaseController
+      def update
+        ::VisitRequest::Cancel.call(visit_request)
+        redirect_to admin_event_visit_requests_path(event)
+      end
+
+      private
+
+      def event
+        @event ||= Event.friendly.find(params[:event_id])
```
Comment (sveredyuk) [inline app/controllers/admin/visit_request/cancel_controller.rb]: DRY!
1. Create BaseController in VisitRequest module
2. extract this method
3. inherit controller from base

File: app/models/visit_request.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class VisitRequest < ApplicationRecord
+  PENDING = 'pending'
```
Comment (sveredyuk) [inline app/models/visit_request.rb]:
```ruby
enum status: %i[PENDING APPROVED CANCELED]
```

But keep symbols in constants

http://edgeapi.rubyonrails.org/classes/ActiveRecord/Enum.html

File: app/services/visit_request/finder/attendee.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+      def call
+        return false unless user
+        return false unless event
+
+        VisitRequest.find_by(user_id: user.id, event_id: event.id)
+      end
+
+      def self.call(*args)
+        new(*args).call
```
Comment (sveredyuk) [inline app/services/visit_request/finder/attendee.rb]: you don't need this method

File: app/views/admin/events/index.slim [hunk 4]
```diff
@@ -20,3 +21,4 @@ table.table
         td = format_timestamp(event.started_at)
         td = format_timestamp(event.finished_at)
         td = admin_edit_event_link(event)
+        td = link_to t('visit_requests.plural'), admin_event_visit_requests_path(event)
```
Comment (sveredyuk) [inline app/views/admin/events/index.slim]: to `admin/events_helper.rb`

File: config/routes.rb [hunk 5]
```diff
@@ -4,7 +4,11 @@
   devise_for :users, controllers: { omniauth_callbacks: 'users/omniauth_callbacks', registrations: 'users/registrations' }
 
   #=== MAIN APP =================================
-  resources :events,  only: %i[index show]
+  resources :events,  only: %i[index show] do
+    resources :visit_requests, controller: 'visit_request/visit_requests', only: %i[create] do
+      patch :cancel, controller: 'visit_request/cancel', action: :update
```
Comment (sveredyuk) [inline config/routes.rb]: Shorter version:
`patch :cancel, to: 'visit_request/cancel#update'`

File: config/routes.rb [hunk 6]
```diff
@@ -18,7 +22,12 @@
   authenticate :user, ->(u) { u.admin? } do
     namespace :admin do
       get '/', to: 'home#index'
-      resources :events,  except: %i[show destroy]
+      resources :events,  except: %i[show destroy] do
+        resources :visit_requests, controller: 'visit_request/visit_requests', only: %i[index] do
+          patch :approve, controller: 'visit_request/approve', action: :update
+          patch :cancel, controller: 'visit_request/cancel', action: :update
```
Comment (sveredyuk) [inline config/routes.rb]: patch :approve, to: 'visit_request/approve#update'
patch :cancel, to: 'visit_request/cancel#update'

File: config/locales/en.yml [hunk 7]
```diff
@@ -79,3 +79,15 @@ en:
     last_name: Last Name
     synthetic: Synthetic
     change_gravatar: Change gravatar
+
+
+  visit_requests:
+    create: 'Attend'
+    create: 'Attend'
```
Comment (sveredyuk) [inline config/locales/en.yml]: remove dup please

File: spec/features/visit_requests/create_spec.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+  context 'when user is not logged in' do
+    before do
+      visit "/events/#{event.id}"
+    end
+
+    it { expect(page).to_not have_link I18n.t('visit_requests.create') }
+    it { expect(page).to have_content I18n.t('visit_requests.unauthorized') }
+  end
+
```
Comment (sveredyuk) [inline spec/features/visit_requests/create_spec.rb]: Why you always add this empty space before `end` ?
I think it's unnecessary

File: app/views/admin/visit_request/visit_requests/index.slim [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+  thead
+    tr
+      th = t 'words.name'
+      th
+      th
+
+  tbody
+    - visit_requests.each do |visit_request|
+      tr
+        th = visit_request.user.full_name
```
Comment (sveredyuk) [inline app/views/admin/visit_request/visit_requests/index.slim]: this also must be a link to member profile with target: _blank

File: app/views/admin/visit_request/visit_requests/index.slim [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+        th = link_to t('visit_requests.approve'), admin_event_visit_request_approve_path(event, visit_request), method: :patch
+        th = link_to t('visit_requests.cancel'), admin_event_visit_request_cancel_path(event, visit_request), method: :patch
```
Comment (sveredyuk) [inline app/views/admin/visit_request/visit_requests/index.slim]: Please, extract all link_to to admin/events_helper and keep template skinny

File: app/views/events/show.slim [hunk 11]
```diff
@@ -1 +1,13 @@
 h2 = t 'events.singular'
+h3 = event.title
+
+- if visit_request
```
Comment (sveredyuk) [inline app/views/events/show.slim:4]: same... all to helpers

**Review by sveredyuk — CHANGES_REQUESTED** [review]
@denysmedynskyi Please, pay attention to all my comments

**Comment by sveredyuk** [comment]
@denysmedynskyi rebase with development branch, new SimpleCov config may help

**Comment by denys-medynskyi** [comment]
@sveredyuk please review again

File: app/views/events/show.slim (hunk 11 above)
Comment (denys-medynskyi) [reply app/views/events/show.slim:4]: this is fixed too

File: app/controllers/visit_request/concerns/visit_request_finder.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+class VisitRequest
```
Comment (sveredyuk) [inline app/controllers/visit_request/concerns/visit_request_finder.rb]: concerns are totally bad idea in rails, it's same as callbacks.

If you want encapsulate get logic from controller, better way will be add new layer - finders.

```ruby
VisitRquest::Find.call()
VisitRquest::FindActive.call()
VisitRquest::FindWhatEverYouNeed.call()
```

File: app/services/visit_request/finder/attendee.rb [hunk 13]
```diff
@@ -0,0 +1,19 @@
+class VisitRequest
+  module Finder
```
Comment (sveredyuk) [inline app/services/visit_request/finder/attendee.rb]: Sorry, but we need do not mix `services` (which represent a process of changes) with `finders`

File: config/routes.rb [hunk 14]
```diff
@@ -4,7 +4,12 @@
   devise_for :users, controllers: { omniauth_callbacks: 'users/omniauth_callbacks', registrations: 'users/registrations' }
 
   #=== MAIN APP =================================
-  resources :events,  only: %i[index show]
+  resources :events,  only: %i[index show] do
+    resources :visit_requests, only: [] do
+      patch :cancel, to: 'visit_request/cancel#update'
```
Comment (sveredyuk) [inline config/routes.rb]: patch is out-of-date REST method. I prefer to use PUT.

**Review by sveredyuk — APPROVED** [review]

File: config/routes.rb (hunk 14 above)
Comment (denys-medynskyi) [reply config/routes.rb]: done

File: app/services/visit_request/finder/attendee.rb (hunk 13 above)
Comment (denys-medynskyi) [reply app/services/visit_request/finder/attendee.rb]: done

**Comment by denys-medynskyi** [comment]
@sveredyuk thanks for suggestions

File: app/finders/visit_request/find.rb [hunk 15]
```diff
@@ -0,0 +1,13 @@
+class VisitRequest
+  class Find < ApplicationService
```
Comment (sveredyuk) [inline app/finders/visit_request/find.rb]: To keep interface segregation  we need `ApplicationFinder` as root for all finders...

File: app/finders/visit_request/find_not_canceled.rb [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+ class FindNotCanceled < ApplicationService
+    attr_reader :event, :user
+
+    def initialize(user, event)
+      @user  = user
+      @event = event
+    end
+
+    def call
+      return unless user
```
Comment (sveredyuk) [inline app/finders/visit_request/find_not_canceled.rb]: ```return unless user || event```

File: app/services/visit_request/attend.rb [hunk 17]
```diff
@@ -0,0 +1,18 @@
+class VisitRequest
+  class Attend < ApplicationService
+
+    def initialize(user, event)
+      @user = user
+      @event = event
+    end
+
+    def call
+      visit_request = VisitRequest.find_or_initialize_by(user: user, event: event)
```
Comment (sveredyuk) [inline app/services/visit_request/attend.rb:10]: user or event can be nil ?

**Review by sveredyuk — APPROVED** [review]

File: app/services/visit_request/attend.rb (hunk 17 above)
Comment (denys-medynskyi) [reply app/services/visit_request/attend.rb:10]: no. `before_authenticate!` filter and `Event.find` verify that

File: app/finders/visit_request/find_not_canceled.rb (hunk 16 above)
Comment (denys-medynskyi) [reply app/finders/visit_request/find_not_canceled.rb]: `return unless user && event` will work as I need. Thanks for suggestion

File: app/finders/visit_request/find.rb (hunk 15 above)
Comment (denys-medynskyi) [reply app/finders/visit_request/find.rb]: ok, also it should not inherit from `ApplicationService`, yes ?

File: app/services/visit_request/attend.rb (hunk 17 above)
Comment (denys-medynskyi) [reply app/services/visit_request/attend.rb:10]: oh, we are talking about service. what do you suggest ?

**Comment by denys-medynskyi** [comment]
@sveredyuk any suggestions ?

**Comment by sveredyuk** [comment]
@denysmedynskyi good job. rebase, squash and merge

**Pull request merged**